ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,138 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
 
 namespace
 {
-    Shape evaluate_matmul_output_shape(const Shape& arg0_shape,
-                                       const Shape& arg1_shape,
-                                       bool transpose_a,
-                                       bool transpose_b)
+    PartialShape validate_matmul_output_shape(const PartialShape& arg0_shape,
+                                              const PartialShape& arg1_shape,
+                                              bool transpose_a,
+                                              bool transpose_b)
     {
-        Shape output_shape;
-        Shape arg0_shape_update = arg0_shape;
-        Shape arg1_shape_update = arg1_shape;
+        auto arg0_rank = arg0_shape.rank().get_length();
+        auto arg1_rank = arg1_shape.rank().get_length();
+        auto max_rank = std::max(arg0_rank, arg1_rank);
+        auto merged_dimension = Dimension::dynamic();
 
-        size_t arg0_rank = arg0_shape.size();
-        size_t arg1_rank = arg1_shape.size();
+        std::vector<Dimension> arg0_shape_update(arg0_rank);
+        std::vector<Dimension> arg1_shape_update(arg1_rank);
+        std::vector<Dimension> output_shape(max_rank);
 
-        if (transpose_a && arg0_rank > 1)
+        for (size_t i = 0; i < arg0_rank; i++)
+            arg0_shape_update[i] = arg0_shape[i];
+        for (size_t i = 0; i < arg1_rank; i++)
+            arg1_shape_update[i] = arg1_shape[i];
```

Comment:
Can we simplify the remaining code by unsqueezing input shapes up to max_rank lenght by adding 1s in the outer dimensions?

Reply:
It can be simplified, but validation for 1D vector has to be preserved. The problem to address here is 1D vector handling.  Assume two scenarios:
1. A[5] x B[5].  Matmul (for  onnx or numpy) will produce scalar output.  In our case after promotion: A[1,5] x B[1, 5] -> it will fail due to dimension mismatch.
2. similar effect for A[1, 2, 5] x B[5]. Onnx or numpy will produce 1D vector whereas an openvino will fail.
It needs to be specified how to handle those scenarios.

Reply:
Not sure I understand. OpenVINO/ONNX make this unsqueezing differently (and ONNX implies some post-squeezing) but it still can be handled before doing checks. BTW, please refer to the e-mail thread, we are tending to change the spec to be aligned with ONNX. But this comment remains relevant to ONNX semantics as well.